File: Assets/Scripts/testing/TromboneEffect.js
var target: Transform;
var VELOCIDAD : float;

private var initHeightAtDist: float;
private var dzEnabled: boolean;


// Calculate the frustum height at a given distance from the camera.
function FrustumHeightAtDistance(distance: float) {
	return 2.0 * distance * Mathf.Tan(GetComponent.<Camera>().fieldOfView * 0.5 * Mathf.Deg2Rad);
}


// Calculate the FOV needed to get a given frustum height at a given distance.
function FOVForHeightAndDistance(height: float, distance: float) {
	return 2 * Mathf.Atan(height * 0.5 / distance) * Mathf.Rad2Deg;
}


// Start the dolly zoom effect.
function StartDZ() {
	var distance = Vector3.Distance(transform.position, target.position);
	initHeightAtDist = FrustumHeightAtDistance(distance);
	dzEnabled = true;
}


// Turn dolly zoom off.
function StopDZ() {
	dzEnabled = false;
}


function Start() {
	StartDZ();
}


function Update () {
	if (dzEnabled) {
		// Measure the new distance and readjust the FOV accordingly.
		var currDistance = Vector3.Distance(transform.position, target.position);
		GetComponent.<Camera>().fieldOfView = FOVForHeightAndDistance(initHeightAtDist, currDistance);
	}

	// Simple control to allow the camera to be moved in and out using the up/down arrows.
	var caca = 0;
	if (Input.GetKey(KeyCode.W))
		caca += 1;
	if (Input.GetKey(KeyCode.S))
		caca -= 1;
	transform.Translate(caca * Vector3.forward * Time.deltaTime * VELOCIDAD);
}

File: Assets/IMPORTED/Scripts/Camara/TromboneEffect.cs
using UnityEngine;
using System.Collections;

[RequireComponent(typeof(Camera))]
public class TromboneEffect : MonoBehaviour {
	
	public Transform target;
	public bool useTrombone = false;
	
	private float _initHeightAtDist;
	private bool _wasUsingTrombone;

	
	// Calculate the frustum height at a given distance from a camera.
	public static float FrustumHeightAtDistance( float distance, Camera cam ) 
	{
		return 2f * distance * Mathf.Tan(cam.fieldOfView * 0.5f * Mathf.Deg2Rad);
	}
	
	// Calculate the FOV needed to get a given frustum height at a given distance.
	public static float FOVForHeightAndDistance( float height, float distance ) 
	{
		return 2f * Mathf.Atan(height * 0.5f / distance) * Mathf.Rad2Deg;
	}
			
	
	void Update () 
	{
		// Should the effect start?
		if ( !_wasUsingTrombone && useTrombone )
		{
			float distance = Vector3.Distance( transform.position, target.position );
			_initHeightAtDist = FrustumHeightAtDistance( distance, GetComponent<Camera>() );
		}
		
		if (useTrombone) 
		{
			// Measure the new distance and readjust the FOV accordingly.
			float currDistance = Vector3.Distance( transform.position, target.position );
			GetComponent<Camera>().fieldOfView = FOVForHeightAndDistance( _initHeightAtDist, currDistance );
		}
		
		_wasUsingTrombone = useTrombone;
	}
	
}

